### PR TITLE
Hassle-free merging of LoRA weights

### LINKED ISSUE #483: Save the model config in the training scripts
while testing conversion of LoRA finetuned checkpoints to their original format, it was found that having a `lora_config.json` saved to `out/lora/` would be beneficial when creating a new model with `model = GPT.from_name(model_name, **lora_config_during_finetuning)` or `model = GPT(**lora_config_during_finetuning)`.

intended changes in `finetune.lora.py` are:

```python
import json
from dataclasses import asdict

def save_lora_config(fabric, model, out_dir: Path):
    fabric.print(f"Saving LoRA config to {str(out_dir)!r}")
    config_payload = {k: v for k, v in asdict(model.config).items() if k != "org"}
    with open(out_dir / "lit_lora_config.json", "w") as config_path:
        json.dump({"config": config_payload}, config_path, indent=4))
```

example output

```json
{
    "config": {
        "name": "Llama-2-7b-hf",
        "block_size": 1,
        "vocab_size": 1,
        "padding_multiple": 64,
        "padded_vocab_size": 64,
        "n_layer": 1,
        "n_head": 4,
        "n_embd": 8,
        "rotary_percentage": 1.0,
        "parallel_residual": false,
        "bias": false,
        "n_query_groups": 4,
        "shared_attention_norm": false,
        "_norm_class": "RMSNorm",
        "norm_eps": 1e-05,
        "_mlp_class": "LLaMAMLP",
        "intermediate_size": 11008,
        "condense_ratio": 1,
        "r": 1,
        "alpha": 8,
        "dropout": 0.1,
        "to_query": true,
        "to_key": false,
        "to_value": true,
        "to_projection": false,
        "to_mlp": false,
        "to_head": false
    }
}
```

### LINKED ISSUE #966: Change `merge_lora.py` output path name
This is a suggestion for making the documentation and user experience to make it more finetuning-script agnostic.

So, currently, 

- `finetune/lora.py` writes a `.../lit_model_lora_finetuned.pth` file 
- `finetune/full.py` writes a `.../lit_model_finetuned.pth` file
- `scripts/merge_lora.py` writes a `.../lit_model.pth` file.

If we want to simplify the documentation and user experience to make it more finetuning-script agnostic, I think it would make sense to have 

- `scripts/merge_lora.py` write a `.../lit_model_finetuned.pth` file.

---

For example, right now:

1. Finetune

```bash
export finetuned_dir=out/lit-finetuned-model

python finetune/lora.py \
   --io.checkpoint_dir checkpoints/$repo_id \
   --io.train_data_dir data/alpaca \
   --io.val_data_dir data/alpaca \
   --train.epochs 1 \
   --io.out_dir $finetuned_dir
```

2. Merge LoRA weights:

Note that this step only applies if the model was finetuned with `lora.py` above and not when `full.py` was used for finetuning.

```bash
python scripts/merge_lora.py \
    --checkpoint_dir checkpoints/$repo_id \
    --lora_path $finetuned_dir/lit_model_lora_finetuned.pth \
    --out_dir $finetuned_dir/
```

3. Convert the finetuning model back into a HF format:

```bash
python scripts/convert_lit_checkpoint.py \
   --checkpoint_path $finetuned_dir/lit_model.pth \
   --output_path out/hf-tinyllama/converted_model.pth \
   --config_path checkpoints/$repo_id/lit_config.json 
```

(If you used `full.py` instead of `lora.py` to finetune your model, 
replace `$finetuned_dir/lit_model.pth` with `$finetuned_dir/lit_model_finetuned.pth`.)

---

If we make this change, we can simplify step 3 to

3. Convert the finetuning model back into a HF format:

```bash
python scripts/convert_lit_checkpoint.py \
   --checkpoint_path $finetuned_dir/lit_model_finetuned.pth \
   --output_path out/hf-tinyllama/converted_model.pth \
   --config_path checkpoints/$repo_id/lit_config.json 
```

### PR DESCRIPTION
**Pull request by awaelchli**
Fixes #483
Fixes #966 

Basically reimplements `scripts/merge_lora.py` by taking the necessary info directly from the new standardized checkpoint folder.

### Before 

```bash
python scripts/merge_lora.py \
	--lora_path out/lora/final/lit_model.pth \
	--checkpoint_dir checkpoints/meta-llama/Llama-2-7b-hf
    --out_dir out/lora/final/merged/
	--lora_r 16 \
    ...
```
What a load! You had to remember three directories, plus the lora hyperparameters you used for finetuning. Aint nobody got time for dis!

### After
Simply pass in the checkpoint you want to merge and the rest is automated. 
```bash
python scripts/merge_lora.py --checkpoint_dir out/lora/final/
```
```
Saved merged weights to 'out/lora/final/lit_model.pth'
A backup of the old LoRA weights is in 'out/lora/final/lit_model.pth.lora'
```

In a follow-up, we will also be able to reuse the logic added in this PR to merge lora-adapters on the fly if merging wasn't done manually prior.

### CODE REVIEW

File: tutorials/finetune_lora.md
Comment (awaelchli) [inline tutorials/finetune_lora.md:1]: Please note how much this simplifies the docs

File: finetune/lora.py [hunk 1]
```diff
@@ -109,15 +109,15 @@ def setup(
         seed,
         Config.from_name(
             name=checkpoint_dir.name,
-            r=lora_r,
```
Comment (awaelchli) [inline finetune/lora.py:112]: The renaming was needed to simplify the logic we have internally to read the config from the saved hyperparameters. Previously, the arguments for the config were named differently than the arguments for the lora.py script.

File: scripts/merge_lora.py [hunk 2]
```diff
@@ -18,66 +20,75 @@
...
-    lora_projection: bool = False,
-    lora_mlp: bool = False,
-    lora_head: bool = False,
 ) -> None:
-    """Generates a response based on a given instruction and an optional input.
-    This script will only work with checkpoints from the instruction-tuned GPT-LoRA model.
-    See `finetune/lora.py`.
+    """Merges the LoRA weights with the base model. See `finetune/lora.py`.
+
+    Merging happens in-place in the checkpoint directory that is given as input.
```
Comment (carmocca) [inline scripts/merge_lora.py:29]: Can you mention here that a copy of the lora weights is generated (and where to)

File: scripts/merge_lora.py [hunk 3]
```diff
@@ -18,66 +20,75 @@
...
+    shutil.move(checkpoint_dir / "lit_model.pth.merged", checkpoint_dir / "lit_model.pth")
+
+    fabric.print(f"Saved merged weights to {str(checkpoint_dir / 'lit_model.pth')!r}")
+    fabric.print(f"A backup of the old LoRA weights is in {str(checkpoint_dir / 'lit_model.pth.lora')!r}")
+
+
+def load_lora_metadata(checkpoint_dir: Path) -> Tuple[Dict[str, Any], Path, str]:
+    hparams_file = checkpoint_dir / "hyperparameters.yaml"
+    if not hparams_file.is_file():
+        raise FileNotFoundError()  # TODO
```
Comment (carmocca) [inline scripts/merge_lora.py:82]: TODO

File: scripts/merge_lora.py [hunk 4]
```diff
@@ -18,66 +20,75 @@
...
 
+    # Make a backup of the LoRA weights (they are only a few MBs)
+    shutil.move(checkpoint_dir / "lit_model.pth", checkpoint_dir / "lit_model.pth.lora")
+    shutil.move(checkpoint_dir / "lit_model.pth.merged", checkpoint_dir / "lit_model.pth")
+
+    fabric.print(f"Saved merged weights to {str(checkpoint_dir / 'lit_model.pth')!r}")
+    fabric.print(f"A backup of the old LoRA weights is in {str(checkpoint_dir / 'lit_model.pth.lora')!r}")
+
+
+def load_lora_metadata(checkpoint_dir: Path) -> Tuple[Dict[str, Any], Path, str]:
```
Comment (carmocca) [inline scripts/merge_lora.py:79]: Technically precision is an Optional[str] since you are using `get`

**Review by carmocca — APPROVED** [review]

File: scripts/merge_lora.py (hunk 4 above)
Comment (awaelchli) [reply scripts/merge_lora.py]: How do you see this stuff 😄

Comment (carmocca) [reply scripts/merge_lora.py]: I installed mypy into my neurocortex

**Pull request merged**